// @dynamic-labs/turnkey/src/lib/TurnkeyPasskeyRecoveryHandler/TurnkeyPasskeyRecoveryHandler.js
import { __awaiter } from '../../../_virtual/_tslib.js';
import { IframeStamper } from '@turnkey/iframe-stamper';
import { TurnkeyClient } from '@turnkey/http';
import { DynamicError } from '@dynamic-labs/utils';
import { logger } from '@dynamic-labs/wallet-connector-core';

const turnkeyBaseUrl = 'https://api.turnkey.com';
const turnkeyRecoveryUrl = 'https://recovery.turnkey.com';
class TurnkeyPasskeyRecoveryHandler {
    get publicKey() {
        return this.__publicKey;
    }
    set recoveryUserId(turnkeyRecoveryUserId) {
        this.__turnkeyRecoveryUserId = turnkeyRecoveryUserId;
    }
    clear() {
        var _a;
        (_a = this.__iframeStamper) === null || _a === void 0 ? void 0 : _a.clear();
        this.__iframeStamper = undefined;
        this.__publicKey = undefined;
        this.__client = undefined;
        this.__turnkeyRecoveryUserId = undefined;
    }
    initRecovery(iframeContainerId, iframeElementId) {
        return __awaiter(this, void 0, void 0, function* () {
            this.__iframeStamper = new IframeStamper({
                iframeContainerId,
                iframeElementId,
                iframeUrl: turnkeyRecoveryUrl,
            });
            yield this.__iframeStamper.init();
            this.__publicKey = this.__iframeStamper.publicKey();
            return this.__publicKey;
        });
    }
    verifyRecoveryCode(recoveryBundle) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.__iframeStamper) {
                throw new DynamicError('Cannot proceed with your request');
            }
            try {
                yield this.__iframeStamper.injectRecoveryBundle(recoveryBundle);
                this.__client = new TurnkeyClient({
                    baseUrl: turnkeyBaseUrl,
                }, this.__iframeStamper);
            }
            catch (err) {
                logger.error('Error while verifying recovery code', err);
                throw err;
            }
        });
    }
    completeRecovery({ attestation, challenge, organizationId, }) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.__client || !this.__turnkeyRecoveryUserId) {
                throw new DynamicError('Cannot proceed with your request');
            }
            try {
                return this.__client.recoverUser({
                    organizationId,
                    parameters: {
                        authenticator: {
                            attestation: attestation,
                            authenticatorName: 'Passkey',
                            challenge,
                        },
                        userId: this.__turnkeyRecoveryUserId,
                    },
                    timestampMs: String(Date.now()),
                    type: 'ACTIVITY_TYPE_RECOVER_USER',
                });
            }
            catch (err) {
                logger.error('Error while completing recovery process', err);
                throw err;
            }
        });
    }
}

export { TurnkeyPasskeyRecoveryHandler };
