// @dynamic-labs/sdk-react-core/src/lib/views/Passkey/PasskeyIntroView/PasskeyIntroView.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../../../_virtual/_tslib.cjs');
var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var reactI18next = require('react-i18next');
var utils = require('@dynamic-labs/utils');
var useInternalDynamicContext = require('../../../context/DynamicContext/useInternalDynamicContext.cjs');
var Typography = require('../../../components/Typography/Typography.cjs');
require('@dynamic-labs/sdk-api');
var logger = require('../../../shared/logger.cjs');
require('@dynamic-labs/iconic');
require('@dynamic-labs/wallet-connector-core');
require('../../../context/ViewContext/ViewContext.cjs');
var questionMark = require('../../../shared/assets/question-mark.cjs');
var passkeyIntroIcon = require('../../../shared/assets/passkey-intro-icon.cjs');
require('@dynamic-labs/wallet-book');
var localStorage$1 = require('../../../utils/constants/localStorage.cjs');
require('../../../utils/constants/colors.cjs');
var localStorage = require('../../../shared/utils/classes/storage/localStorage.cjs');
require('viem');
require('../../../shared/consts/index.cjs');
var index$1 = require('../../../utils/functions/storeAuthToken/index.cjs');
require('@dynamic-labs/multi-wallet');
var createUserProfile = require('../../../utils/functions/createUserProfile/createUserProfile.cjs');
require('../../../components/ShadowDOM/ShadowDOM.cjs');
require('../../../context/ThemeContext/ThemeContext.cjs');
require('react-dom');
require('react-focus-lock');
require('../../../components/Transition/ZoomTransition/ZoomTransition.cjs');
require('../../../components/Transition/SlideInUpTransition/SlideInUpTransition.cjs');
require('../../../components/Transition/OpacityTransition/OpacityTransition.cjs');
require('../../../config/ApiEndpoint.cjs');
require('@dynamic-labs/rpc-providers');
require('../../../utils/hooks/useWallets/utils/verifyMagicIsSetup/verifyMagicIsSetup.cjs');
require('../../../context/ErrorContext/ErrorContext.cjs');
require('../../../context/AccessDeniedContext/AccessDeniedContext.cjs');
require('../../../context/AccountExistsContext/AccountExistsContext.cjs');
require('../../../components/OverlayCard/OverlayCard.context.cjs');
var index = require('../../../context/FooterAnimationContext/index.cjs');
require('../../../context/MockContext/MockContext.cjs');
require('../../../context/QrCodeContext/QrCodeContext.cjs');
var Tooltip = require('../../../components/Tooltip/Tooltip.cjs');
require('../../../context/WalletGroupContext/WalletGroupContext.cjs');
require('../../../widgets/DynamicWidget/components/DynamicWidgetHeader/DynamicWidgetHeader.cjs');
var usePromise = require('../../../utils/hooks/usePromise/usePromise.cjs');
require('../../../context/CaptchaContext/CaptchaContext.cjs');
require('../../../context/EmailVerificationContext/EmailVerificationContext.cjs');
require('qrcode');
require('../../../context/LoadingContext/LoadingContext.cjs');
require('i18next');
require('../../../widgets/DynamicWidget/context/DynamicWidgetContext.cjs');
require('yup');
var ErrorContainer = require('../../../components/ErrorContainer/ErrorContainer.cjs');
require('../../../components/UserProfile/parts/UserProfileField/components/VerifiedEmailIcon/VerifiedEmailIcon.cjs');
require('@dynamic-labs/types');
require('../../../utils/hooks/useUserUpdateRequest/useUpdateUser/useUpdateUser.cjs');
require('../../../context/UserFieldEditorContext/UserFieldEditorContext.cjs');
require('formik');
var Button = require('../../../components/Button/Button.cjs');
var ModalHeader = require('../../../components/ModalHeader/ModalHeader.cjs');
var IconButton = require('../../../components/IconButton/IconButton.cjs');
require('../../../components/Alert/Alert.cjs');
require('../../../widgets/DynamicWidget/views/ManagePasskeysWidgetView/PasskeyCard/PasskeyCard.cjs');
var getProperErrorMessage = require('../../../modals/SignMessageConfirmationModal/getProperErrorMessage.cjs');
var useMutation = require('../../../utils/hooks/useMutation/useMutation.cjs');
var TextButton = require('../../../components/TextButton/TextButton.cjs');
require('../../../components/Popper/Popper/Popper.cjs');
require('../../../components/Popper/PopperContext/PopperContext.cjs');
require('../../../components/InlineWidget/InlineWidget.cjs');
require('../../../components/MenuList/Dropdown/Dropdown.cjs');
var isPasskeyEnabled = require('../../../utils/functions/isPasskeyEnabled/isPasskeyEnabled.cjs');
var createEmbeddedWalletWithPasskey = require('../utils/createEmbeddedWalletWithPasskey/createEmbeddedWalletWithPasskey.cjs');
var findPasskeyEmailWalletConnector = require('../utils/findPasskeyEmailWalletConnector/findPasskeyEmailWalletConnector.cjs');
var PasskeyDeviceIcon = require('../PasskeyDeviceIcon/PasskeyDeviceIcon.cjs');

const PasskeyIntroView = ({ createEmbeddedWalletMutation = createEmbeddedWalletWithPasskey.createEmbeddedWalletWithPasskey, }) => {
    const { user, wallets, setPrimaryWalletId, setUser, authToken, environmentId, handleLogOut, internalEvents, setShowAuthFlow, setCallback, } = useInternalDynamicContext.useInternalDynamicContext();
    const { setIsFooterExpanded } = index.useFooterAnimationContext();
    const { t } = reactI18next.useTranslation();
    const { data: passkeyEnabled } = usePromise.usePromise(isPasskeyEnabled.isPasskeyEnabled);
    // if passkey is not yet setup, but user logged in with email in a passkey env
    // it should be null to force user to setup passkey even after the a refresh
    React.useEffect(() => localStorage.LocalStorage.removeFromLS(localStorage$1.LAST_USED_WALLET), []);
    const { mutate: createEmbeddedWalletWithPasskeyMutation, isLoading, error, } = useMutation.useMutation(() => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        var _a;
        const { embeddedWalletVerifiedCredentialId, jwt, decodedJwt, walletConnector, } = yield createEmbeddedWalletMutation({
            authToken,
            environmentId,
            user,
            walletConnector: findPasskeyEmailWalletConnector.findPasskeyEmailWalletConnector(wallets),
        });
        const passkeyWallet = {
            address: (yield walletConnector.fetchPublicAddress()) || '',
            chain: walletConnector.connectedChain || '',
            connected: true,
            connector: walletConnector,
            id: embeddedWalletVerifiedCredentialId,
        };
        internalEvents.current.emit('embeddedWalletCreated', passkeyWallet);
        setPrimaryWalletId(embeddedWalletVerifiedCredentialId);
        index$1.storeAuthToken(jwt);
        setUser(createUserProfile.createUserProfile(decodedJwt));
        setShowAuthFlow(false);
        const turnkeyVerifiedCredentials = (_a = decodedJwt === null || decodedJwt === void 0 ? void 0 : decodedJwt.verifiedCredentials) === null || _a === void 0 ? void 0 : _a.find(({ walletName }) => walletName === 'turnkey');
        setCallback('embeddedWalletCreated', {
            email: (turnkeyVerifiedCredentials === null || turnkeyVerifiedCredentials === void 0 ? void 0 : turnkeyVerifiedCredentials.email) || '',
            verifiedCredential: turnkeyVerifiedCredentials,
        });
    }), {
        onFailure: (err) => {
            logger.logger.error('Failed to create embedded wallet', err);
            internalEvents.current.emit('embeddedWalletFailed', err);
        },
    });
    const errorText = React.useMemo(() => {
        if (!error) {
            return undefined;
        }
        if (error instanceof utils.DynamicError) {
            return error.message;
        }
        return getProperErrorMessage.getProperErrorMessage(error);
    }, [error]);
    const infoIconButton = (jsxRuntime.jsx(Tooltip.Tooltip, { content: t('dyn_passkey_intro.helper.tooltip'), className: 'header__tooltip', children: jsxRuntime.jsx(IconButton.IconButton, { type: 'button', "data-testid": 'info-button', onClick: () => setIsFooterExpanded((prev) => !prev), className: 'header__icon', children: jsxRuntime.jsx(questionMark.ReactComponent, {}) }) }, 'info-button'));
    return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx(ModalHeader.ModalHeader, { trailing: infoIconButton, children: jsxRuntime.jsx(Typography.Typography, { as: 'h1', variant: 'title', color: 'primary', "data-testid": 'dynamic-auth-modal-heading', className: 'header__typography', children: t('dyn_passkey_intro.title') }) }), jsxRuntime.jsxs("div", { className: 'passkey-intro-view', children: [jsxRuntime.jsx("div", { className: 'passkey-intro-view__header', children: jsxRuntime.jsx(passkeyIntroIcon.ReactComponent, {}) }), jsxRuntime.jsx(Typography.Typography, { variant: 'body_normal', weight: 'regular', color: 'secondary', className: 'passkey-intro-view__subtitle', children: t('dyn_passkey_intro.subtitle') }), passkeyEnabled === false && (jsxRuntime.jsx(ErrorContainer.ErrorContainer, { children: t('dyn_passkey_intro.disabled') })), errorText && jsxRuntime.jsx(ErrorContainer.ErrorContainer, { children: errorText }), jsxRuntime.jsxs("div", { className: 'passkey-intro-view__actions', children: [jsxRuntime.jsx(Button.Button, { buttonVariant: 'brand-primary', buttonPadding: 'large', onClick: () => createEmbeddedWalletWithPasskeyMutation(), dataTestId: 'setup-passkey-button', loading: isLoading, disabled: isLoading || !passkeyEnabled, typographyProps: {
                                    color: 'inherit',
                                }, children: jsxRuntime.jsxs("div", { className: 'passkey-intro-view__inline-button', children: [jsxRuntime.jsx(PasskeyDeviceIcon.PasskeyDeviceIcon, {}), jsxRuntime.jsx(Typography.Typography, { children: t('dyn_passkey_intro.button') })] }) }), jsxRuntime.jsx(TextButton.TextButton, { "data-testid": 'setup-passkey-logout-button', className: 'passkey-intro-view__log-out', onClick: handleLogOut, children: t('dyn_passkey_intro.button_logout') })] })] })] }));
};

exports.PasskeyIntroView = PasskeyIntroView;
