// @dynamic-labs/wallet-book/src/helpers/getWalletLinks.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var findWalletBookWallet = require('./findWalletBookWallet.cjs');
var renderTemplate = require('./renderTemplate.cjs');

const getWalletLinks = (walletBook, walletKey) => {
    const walletData = findWalletBookWallet.findWalletBookWallet(walletBook, walletKey);
    const links = {
        android: '',
        brave: '',
        chrome: '',
        edge: '',
        firefox: '',
        ios: '',
    };
    links.brave =
        renderTemplate.renderTemplate('chromeUrl', walletData?.desktop?.chromeId) ?? '';
    links.chrome =
        renderTemplate.renderTemplate('chromeUrl', walletData?.desktop?.chromeId) ?? '';
    links.edge = renderTemplate.renderTemplate('edgeUrl', walletData?.desktop?.edgeId) ?? '';
    links.firefox =
        renderTemplate.renderTemplate('firefoxUrl', walletData?.desktop?.firefoxId) ?? '';
    links.ios =
        renderTemplate.renderTemplate('iosUrl', walletData?.mobile?.iosId) ??
            walletData?.mobile?.ios ??
            '';
    links.android =
        renderTemplate.renderTemplate('androidUrl', walletData?.mobile?.androidId) ??
            walletData?.mobile?.android ??
            '';
    return links;
};

exports.getWalletLinks = getWalletLinks;
