// @dynamic-labs/sdk-react-core/src/lib/views/SendBalanceView/SendBalanceView.js
import { __awaiter } from '../../../../_virtual/_tslib.js';
import { jsx } from 'react/jsx-runtime';
import { useState, useRef } from 'react';
import { useTranslation } from 'react-i18next';
import { parseEther } from 'viem';
import { DynamicError } from '@dynamic-labs/utils';
import { getChainInfo } from '@dynamic-labs/wallet-connector-core';
import '@dynamic-labs/sdk-api';
import '../../config/ApiEndpoint.js';
import '../../shared/logger.js';
import { getChainIcon } from '../../shared/utils/functions/chain/getChainIcon.js';
import '@dynamic-labs/wallet-book';
import { shortenWalletAddress } from '../../shared/utils/functions/shortenWalletAddress/shortenWalletAddress.js';
import '../../utils/constants/colors.js';
import '../../shared/utils/classes/storage/localStorage.js';
import { formatBigNumber } from '../../shared/utils/functions/formatBigNumber/formatBigNumber.js';
import '@dynamic-labs/iconic';
import '../../context/ViewContext/ViewContext.js';
import '../../shared/consts/index.js';
import '@dynamic-labs/multi-wallet';
import '@dynamic-labs/rpc-providers';
import '../../utils/hooks/useWallets/utils/verifyMagicIsSetup/verifyMagicIsSetup.js';
import { useInternalDynamicContext } from '../../context/DynamicContext/useInternalDynamicContext.js';
import '../../context/ErrorContext/ErrorContext.js';
import '../../context/AccessDeniedContext/AccessDeniedContext.js';
import '../../context/AccountExistsContext/AccountExistsContext.js';
import 'react-dom';
import '../../components/ShadowDOM/ShadowDOM.js';
import '../../context/ThemeContext/ThemeContext.js';
import { Icon } from '../../components/Icon/Icon.js';
import 'react-focus-lock';
import '../../components/Transition/ZoomTransition/ZoomTransition.js';
import '../../components/Transition/SlideInUpTransition/SlideInUpTransition.js';
import '../../components/Transition/OpacityTransition/OpacityTransition.js';
import '../../components/IconButton/IconButton.js';
import '../../components/Alert/Alert.js';
import '../../components/Popper/Popper/Popper.js';
import '../../components/Popper/PopperContext/PopperContext.js';
import { SendBalancePageLayout } from '../../components/SendBalancePageLayout/SendBalancePageLayout.js';
import '../../components/InlineWidget/InlineWidget.js';
import '../../components/MenuList/Dropdown/Dropdown.js';
import '../../components/OverlayCard/OverlayCard.context.js';
import '../../context/FooterAnimationContext/index.js';
import '../../context/MockContext/MockContext.js';
import '../../context/QrCodeContext/QrCodeContext.js';
import '../../context/WalletGroupContext/WalletGroupContext.js';
import '../../widgets/DynamicWidget/components/DynamicWidgetHeader/DynamicWidgetHeader.js';
import { usePromise } from '../../utils/hooks/usePromise/usePromise.js';
import '../../context/CaptchaContext/CaptchaContext.js';
import '../../context/EmailVerificationContext/EmailVerificationContext.js';
import 'qrcode';
import '../../context/LoadingContext/LoadingContext.js';
import 'i18next';
import '../../widgets/DynamicWidget/context/DynamicWidgetContext.js';
import 'yup';
import '../../components/UserProfile/parts/UserProfileField/components/VerifiedEmailIcon/VerifiedEmailIcon.js';
import '@dynamic-labs/types';
import '../../utils/hooks/useUserUpdateRequest/useUpdateUser/useUpdateUser.js';
import '../../context/UserFieldEditorContext/UserFieldEditorContext.js';
import 'formik';
import { TransactionStatusLayout } from '../../components/TransactionStatusLayout/TransactionStatusLayout.js';
import { useFetchCurrency } from '../../widgets/DynamicWidget/hooks/useFetchCurrency/useFetchCurrency.js';
import { TransactionStage } from './components/TransactionStage/TransactionStage.js';
import '../../widgets/DynamicWidget/views/ManagePasskeysWidgetView/PasskeyCard/PasskeyCard.js';

const SendBalanceView = ({ initialRecipientAddress = '', initialValue, onClickBack, onClickClose, onError, onSuccess, onDone, displayPoweredByDynamicFooter = false, }) => {
    const [stage, setStage] = useState('form');
    const { primaryWallet, network, walletUiUtils } = useInternalDynamicContext();
    const { t } = useTranslation();
    const [value, setValue] = useState(initialValue !== null && initialValue !== void 0 ? initialValue : null);
    const [recipientAddress, setRecipientAddress] = useState(initialRecipientAddress);
    const transactionRef = useRef(null);
    if (!primaryWallet) {
        throw new DynamicError('Primary wallet is not available, send balance cannot be displayed');
    }
    const walletConnector = primaryWallet.connector;
    const { data: balance } = usePromise(() => walletConnector.getBalance(), {
        deps: [walletConnector],
    });
    const { data: provider } = usePromise(() => primaryWallet.connector.getPublicClient());
    const evmNetworks = (walletConnector && walletConnector.evmNetworks) || [];
    const evmNetwork = evmNetworks.find((evmNetwork) => evmNetwork.chainId === network);
    const networkIcon = evmNetwork && (jsx("img", { src: evmNetwork === null || evmNetwork === void 0 ? void 0 : evmNetwork.iconUrls[0], alt: `${evmNetwork === null || evmNetwork === void 0 ? void 0 : evmNetwork.name} icon` }));
    const networkName = (evmNetwork === null || evmNetwork === void 0 ? void 0 : evmNetwork.vanityName) || (evmNetwork === null || evmNetwork === void 0 ? void 0 : evmNetwork.name);
    const networkCurrencyDecimals = evmNetwork === null || evmNetwork === void 0 ? void 0 : evmNetwork.nativeCurrency.decimals;
    const ChainIconComponent = getChainIcon(primaryWallet.chain);
    const chainInfo = getChainInfo(primaryWallet.chain);
    const chainIcon = (jsx(Icon, { size: 'small', children: jsx(ChainIconComponent, {}) }));
    const [currencySymbol = ''] = useFetchCurrency(primaryWallet.connector);
    const buildFormStage = () => (jsx(SendBalancePageLayout, { initialValues: {
            amount: value === null ? '' : formatBigNumber(value),
            recipient: recipientAddress,
        }, balance: balance, onClickBack: onClickBack, chain: primaryWallet.chain, networkName: networkName, networkIcon: networkIcon, networkCurrencyDecimals: networkCurrencyDecimals, currencySymbol: currencySymbol, isLoading: false, walletAddress: shortenWalletAddress(primaryWallet.address, 3, 3), walletKey: primaryWallet.connector.key, onClickClose: onClickClose, displayPoweredByDynamicFooter: displayPoweredByDynamicFooter, onSubmit: (data) => {
            setValue(parseEther(data.amount, 'wei'));
            setRecipientAddress(data.recipient);
            setStage('confirmation');
        } }));
    const buildTransactionStage = () => {
        if (!provider || value === null) {
            return null;
        }
        const transaction = {
            from: primaryWallet.address,
            to: recipientAddress,
            value,
        };
        return (jsx(TransactionStage, { walletConnector: walletConnector, provider: provider, transaction: transaction, title: t('dyn_send_transaction.confirmation.title'), onClickBack: () => setStage('form'), onError: onError, displayPoweredByDynamicFooter: displayPoweredByDynamicFooter, mutation: ({ to, value, gasPrice, nonce }) => __awaiter(void 0, void 0, void 0, function* () {
                walletUiUtils.disabledConfirmationOnce();
                const signer = yield walletConnector.getSigner();
                const sentTransaction = yield signer.sendTransaction({
                    account: signer.account || primaryWallet.address,
                    /**
                     * chain is set to null because viem already have the correct chain from the wallet connector
                     * docs: https://viem.sh/docs/actions/wallet/sendTransaction.html#chain-optional
                     */
                    chain: null,
                    gasPrice,
                    nonce,
                    to,
                    value,
                });
                return sentTransaction;
            }), onSuccess: (transaction) => {
                transactionRef.current = transaction;
                setStage('success');
                onSuccess === null || onSuccess === void 0 ? void 0 : onSuccess(transactionRef.current);
            } }));
    };
    const buildSuccessState = () => {
        var _a;
        if (value === null)
            return null;
        return (jsx(TransactionStatusLayout, { amount: formatBigNumber(value), destinationAddress: recipientAddress, networkCurrency: currencySymbol, networkName: (_a = chainInfo === null || chainInfo === void 0 ? void 0 : chainInfo.blockchainName) !== null && _a !== void 0 ? _a : '', NetworkIcon: chainIcon, onClickClose: onClickClose, onDone: () => onDone === null || onDone === void 0 ? void 0 : onDone(), displayPoweredByDynamicFooter: displayPoweredByDynamicFooter }));
    };
    const getCurrentView = () => {
        switch (stage) {
            case 'form':
                return buildFormStage();
            case 'confirmation':
                return buildTransactionStage();
            case 'success':
                return buildSuccessState();
        }
    };
    return getCurrentView();
};

export { SendBalanceView };
