// @dynamic-labs/wallet-book/src/schemas/walletSchema.js
import { z, string } from 'zod';
import { nonEmptyString } from './utils/nonEmptyString.js';
import { transformChromeExtensionId } from './utils/transformChromeExtensionId.js';
import { transformEdgeExtensionId } from './utils/transformEdgeExtensionId.js';
import { transformFirefoxExtensionId } from './utils/transformFirefoxExtensionId.js';
import { transformIosId } from './utils/transformIosId.js';
import { transformAndroidId } from './utils/transformAndroidId.js';
import { nonEmptyStringArray } from './utils/nonEmptyStringArray.js';
import { filterEmptyObject } from './utils/filterEmptyObject.js';

const injectedConfigSchema = z.object({
    chain: z.string(),
    extensionLocators: z.array(z.object({
        flag: z.string(),
        value: z.boolean().optional().default(true),
    })),
    windowLocations: z
        .array(z.string())
        .optional()
        .refine((val) => {
        if (!val)
            return true;
        if (!val.some((v) => ['ethereum', 'ethereum.providers'].includes(v)))
            return true;
        return false;
    }, {
        message: 'windowLocations cannot include ethereum or ethereum.providers as they are included by default',
        path: ['config'],
    }),
});
const walletSchema = z
    .preprocess((val) => val, z.object({
    brand: z
        .object({
        alt: nonEmptyString,
        imageId: nonEmptyString,
        primaryColor: nonEmptyString,
        spriteId: nonEmptyString,
    })
        .optional()
        .refine((val) => {
        if (!val)
            return true;
        if (!(val.spriteId && val.imageId) &&
            (val.spriteId || val.imageId)) {
            return true;
        }
        return false;
    }, {
        message: 'Only one of spriteId or imageId can be defined',
        path: ['brand'],
    }),
    chains: z.array(z.string()).optional(),
    desktop: z
        .object({
        chromeId: nonEmptyString.transform(transformChromeExtensionId),
        edgeId: nonEmptyString.transform(transformEdgeExtensionId),
        firefoxId: nonEmptyString.transform(transformFirefoxExtensionId),
        native: nonEmptyString,
        operaId: nonEmptyString,
        safariId: nonEmptyString,
        universal: nonEmptyString,
    })
        .optional()
        .transform(filterEmptyObject),
    filterFromWalletConnect: z.boolean().optional(),
    group: z.string().optional(),
    injectedConfig: z.array(injectedConfigSchema).optional(),
    mobile: z
        .object({
        android: string().nullish(),
        androidId: nonEmptyString.transform(transformAndroidId),
        ios: string().nullish(),
        iosId: nonEmptyString.transform(transformIosId),
        native: nonEmptyString,
        universal: nonEmptyString,
    })
        .optional()
        .transform(filterEmptyObject),
    name: z.string(),
    shortName: nonEmptyString,
    showOnlyIfInstalled: z.boolean().optional(),
    switchNetworkOnlyFromWallet: z.boolean().optional(),
    walletConnect: z
        .object({
        sdks: nonEmptyStringArray,
    })
        .optional()
        .transform(filterEmptyObject),
}))
    .transform((val) => {
    if (val.mobile?.iosId || val.mobile?.ios === null) {
        delete val.mobile?.ios;
    }
    if (val.mobile?.androidId || val.mobile?.android === null) {
        delete val.mobile?.android;
    }
    return val;
});

export { walletSchema };
