// @dynamic-labs/sdk-react-core/src/lib/utils/hooks/useWalletConnectors/utils/createLinkedWalletsFromWalletOptions/createLinkedWalletsFromWalletOptions.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var decodeJwt = require('../../../../../shared/utils/functions/decodeJwt/decodeJwt.cjs');
require('@dynamic-labs/iconic');
require('@dynamic-labs/wallet-connector-core');
require('react/jsx-runtime');
require('../../../../../context/ViewContext/ViewContext.cjs');
require('react');
require('../../../../../shared/logger.cjs');
require('@dynamic-labs/wallet-book');
require('../../../../constants/colors.cjs');
require('../../../../../shared/utils/classes/storage/localStorage.cjs');
require('viem');
require('@dynamic-labs/utils');
require('../../../../../shared/consts/index.cjs');
var isSameWalletName = require('../../../../functions/isSameWalletName/isSameWalletName.cjs');
require('@dynamic-labs/multi-wallet');
require('@dynamic-labs/sdk-api');
var shouldManuallyReconnectOnRefresh = require('../../../../functions/shouldManuallyReconnectOnRefresh/shouldManuallyReconnectOnRefresh.cjs');

const createLinkedWalletsFromWalletOptions = ({ authToken, walletOptions, primaryWalletId, }) => {
    const decodedJwt = decodeJwt.decodeJwt(authToken);
    // wallet state should be null while loading all the wallets or when there is no user
    if (!walletOptions || !decodedJwt) {
        return [];
    }
    return decodedJwt.verifiedCredentials
        .map((account) => {
        const wallet = walletOptions.find((wallet) => account.walletName &&
            isSameWalletName.isSameWalletName(wallet.name, account.walletName));
        // this probably shouldn't happen. this would mean that the user has an account linked
        // with wallet W, but the customer has toggled off wallet W or the chain that
        // supports wallet W
        /* istanbul ignore next */
        if (!wallet)
            return null;
        if (account.id === primaryWalletId &&
            shouldManuallyReconnectOnRefresh.shouldManuallyReconnectOnRefresh(wallet.walletConnector)) {
            wallet.walletConnector.connect();
        }
        if (account.address && account.chain) {
            return {
                address: account.address,
                chain: account.chain,
                connected: false,
                connector: wallet.walletConnector,
                id: account.id,
            };
        }
        return null;
    })
        .filter((wc) => wc !== null);
};

exports.createLinkedWalletsFromWalletOptions = createLinkedWalletsFromWalletOptions;
