// @dynamic-labs/sdk-react-core/src/lib/utils/hooks/useWalletConnectors/utils/createLinkedWalletsFromWalletOptions/createLinkedWalletsFromWalletOptions.js
import { decodeJwt } from '../../../../../shared/utils/functions/decodeJwt/decodeJwt.js';
import '@dynamic-labs/iconic';
import '@dynamic-labs/wallet-connector-core';
import 'react/jsx-runtime';
import '../../../../../context/ViewContext/ViewContext.js';
import 'react';
import '../../../../../shared/logger.js';
import '@dynamic-labs/wallet-book';
import '../../../../constants/colors.js';
import '../../../../../shared/utils/classes/storage/localStorage.js';
import 'viem';
import '@dynamic-labs/utils';
import '../../../../../shared/consts/index.js';
import { isSameWalletName } from '../../../../functions/isSameWalletName/isSameWalletName.js';
import '@dynamic-labs/multi-wallet';
import '@dynamic-labs/sdk-api';
import { shouldManuallyReconnectOnRefresh } from '../../../../functions/shouldManuallyReconnectOnRefresh/shouldManuallyReconnectOnRefresh.js';

const createLinkedWalletsFromWalletOptions = ({ authToken, walletOptions, primaryWalletId, }) => {
    const decodedJwt = decodeJwt(authToken);
    // wallet state should be null while loading all the wallets or when there is no user
    if (!walletOptions || !decodedJwt) {
        return [];
    }
    return decodedJwt.verifiedCredentials
        .map((account) => {
        const wallet = walletOptions.find((wallet) => account.walletName &&
            isSameWalletName(wallet.name, account.walletName));
        // this probably shouldn't happen. this would mean that the user has an account linked
        // with wallet W, but the customer has toggled off wallet W or the chain that
        // supports wallet W
        /* istanbul ignore next */
        if (!wallet)
            return null;
        if (account.id === primaryWalletId &&
            shouldManuallyReconnectOnRefresh(wallet.walletConnector)) {
            wallet.walletConnector.connect();
        }
        if (account.address && account.chain) {
            return {
                address: account.address,
                chain: account.chain,
                connected: false,
                connector: wallet.walletConnector,
                id: account.id,
            };
        }
        return null;
    })
        .filter((wc) => wc !== null);
};

export { createLinkedWalletsFromWalletOptions };
