// @dynamic-labs/sdk-react-core/src/lib/views/WalletList/WalletList.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { useState, useRef } from 'react';
import { useTranslation } from 'react-i18next';
import { useWalletBookContext } from '@dynamic-labs/wallet-book';
import { ErrorContainer } from '../../components/ErrorContainer/ErrorContainer.js';
import { Search } from '../../components/Search/Search.js';
import { Skeleton } from '../../components/Skeleton/Skeleton.js';
import { useErrorContext } from '../../context/ErrorContext/ErrorContext.js';
import { WALLET_PICKER_SEARCH_KEY, LAST_USED_WALLET } from '../../utils/constants/localStorage.js';
import '../../utils/constants/colors.js';
import { walletListBuilder } from '../../utils/functions/walletListBuilder/walletListBuilder.js';
import { classNames } from '../../utils/functions/classNames/classNames.js';
import { useInternalDynamicContext } from '../../context/DynamicContext/useInternalDynamicContext.js';
import { useMockContext } from '../../context/MockContext/MockContext.js';
import { useViewContext } from '../../context/ViewContext/ViewContext.js';
import { isUserLinkingWallet } from '../../shared/utils/functions/isUserLinkingWallet/isUserLinkingWallet.js';
import { FilterWalletsByChainName } from '../../utils/functions/walletFilters/index.js';
import '@dynamic-labs/sdk-api';
import '../../shared/logger.js';
import '@dynamic-labs/iconic';
import '@dynamic-labs/wallet-connector-core';
import { LocalStorage } from '../../shared/utils/classes/storage/localStorage.js';
import 'viem';
import '@dynamic-labs/utils';
import '../../shared/consts/index.js';
import { ReactComponent as SvgNoWalletFound } from '../../assets/no-wallet-found.js';
import { SearchNotFoundMessage } from './SearchNotFoundMessage/SearchNotFoundMessage.js';
import { WalletListItem } from './WalletListItem/WalletListItem.js';

LocalStorage.setToLS(WALLET_PICKER_SEARCH_KEY, '');
const WalletList = ({ isWalletConnectList = false, }) => {
    var _a;
    const { walletsFilter, signWithEmailWalletName, user, wallets, projectSettings, defaultNumberOfWalletsToShow, walletsToConnectByChain, connectedWallets, authMode, multiWallet, } = useInternalDynamicContext();
    const { walletBook } = useWalletBookContext();
    const { error } = useErrorContext();
    const { view } = useViewContext();
    const { mockedSDK } = useMockContext();
    const { t } = useTranslation();
    const [filterValue, setFilterValue] = useState((_a = LocalStorage.getFromLS(WALLET_PICKER_SEARCH_KEY)) !== null && _a !== void 0 ? _a : '');
    const [footerIsVisible, setFooterIsVisibile] = useState(false);
    const getFilteredWalletsByChain = (walletsToConnectByChain === null || walletsToConnectByChain === void 0 ? void 0 : walletsToConnectByChain.length) &&
        FilterWalletsByChainName(walletsToConnectByChain[0].chain);
    const filteredWalletsByChain = getFilteredWalletsByChain && getFilteredWalletsByChain(wallets);
    const { numberOfWallets, walletsList } = walletListBuilder({
        authMode,
        connectedWallets,
        groupWallets: true,
        inputList: filteredWalletsByChain || wallets,
        isWalletConnectList,
        lastUsedWalletName: LocalStorage.getFromLS(LAST_USED_WALLET),
        multiWallet,
        numberOfWalletsToShow: defaultNumberOfWalletsToShow,
        searchFilter: filterValue,
        showMoreWalletsWithFilter: true,
        signWithEmailWalletName,
        walletBook,
        walletsFilter,
    });
    const isSearchEnabled = numberOfWallets > defaultNumberOfWalletsToShow &&
        walletsList.length !== numberOfWallets;
    const noDefaultFooter = isUserLinkingWallet(user, view, mockedSDK);
    const walletListScrollRef = useRef(null);
    /* istanbul ignore next */
    const handleScroll = () => {
        const element = walletListScrollRef.current;
        if (!element) {
            return;
        }
        if ((element === null || element === void 0 ? void 0 : element.scrollTop) > (element === null || element === void 0 ? void 0 : element.clientHeight) * 1.25) {
            setFooterIsVisibile(true);
        }
        else {
            setFooterIsVisibile(false);
        }
    };
    const handleFilterValueChange = (value) => {
        LocalStorage.setToLS(WALLET_PICKER_SEARCH_KEY, value);
        setFilterValue(value);
    };
    const searchContainer = !projectSettings ? (jsx(Skeleton, { className: 'wallet-list__search-skeleton' })) : (isSearchEnabled && (jsx("div", { className: classNames('wallet-list__search-container', {
            'wallet-list__search-container--scroll': !error,
        }), children: jsx(Search, { label: t('dyn_wallet_list.search.label', { numberOfWallets }), value: filterValue, onChange: ({ target: { value } }) => handleFilterValueChange(value), onClickClear: () => handleFilterValueChange('') }) })));
    return (jsxs(Fragment, { children: [searchContainer, Boolean(error) && (jsx(ErrorContainer, { className: 'wallet-list__error-container', withIcon: false, children: error })), jsx("div", { className: 'wallet-list__container', children: jsxs("div", { className: classNames('wallet-list__scroll-container', {
                        'wallet-list__scroll-container--error': Boolean(error),
                        'wallet-list__scroll-container--full-height': isSearchEnabled,
                    }), "data-testid": 'wallet-list-scroll-container', ref: walletListScrollRef, onScroll: handleScroll, children: [!projectSettings ? (jsx(Skeleton, { count: 10, className: 'wallet-list__tile-skeleton' })) : (jsxs(Fragment, { children: [wallets.length === 0 && (jsx(ErrorContainer, { children: t('dyn_wallet_list.configuration_mismatch') })), wallets.length && walletsList.length === 0 ? (jsx(SearchNotFoundMessage, { title: t('dyn_wallet_list.search.not_found.title'), subtitle: t('dyn_wallet_list.search.not_found.description'), image: jsx(SvgNoWalletFound, {}) })) : (walletsList.map((wallet, index) => (jsx(WalletListItem, { wallet: wallet, usingMultiWallet: multiWallet, onResetSearchValue: () => handleFilterValueChange('') }, `${wallet.key}_${index}`))))] })), isSearchEnabled && !filterValue && projectSettings && (jsx(SearchNotFoundMessage, {}))] }) }), noDefaultFooter && (jsx("div", { className: classNames('wallet-list__tiny-footer', {
                    'wallet-list__tiny-footer--hidden': footerIsVisible,
                }), "data-testid": 'wallet-list-tiny-footer' }))] }));
};

export { WalletList };
